contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+PV as a source of truth. All these external components will also need to be
+updated to look at the PV Annotation to determine whether it is meant for that
+CSI Driver or not, and not just at the PV Source. 
+
+For Node calls, we can put the CSI Source JSON into an annotation on the
+in-memory PV in the VolumeToMount object. This is enough since the information
+in this object is propagated all the way through to the CSI components when
+SetUp is called.
+
+FindPluginBySpec will also have to be modified to return the CSI plugin if the
+plugin migration flag is set for this particular plugin and the annotation has
+been added.
+
+Then the rest of the code in the Operation Generator can execute as normal with
+the CSI Plugin and the annotation in the requisite locations.
+
+GetCSISourceFromSpec is used by both Node/Controller in-tree CSI functions and
+can be changed to get the CSI Source from the annotation
+PV.Annotations[‘k8s.io/CSIVolumeSource‘] first if it exists. This will ensure
+all the in-tree CSI plugin functions will be using the CSI Source translated
+from before.
+
+Caveat: For ALL detach calls of plugins that MAY have already been migrated we
+have to attempt to DELETE the VolumeAttachment object that would have been
+created if that plugin was migrated. This is because Attach after migration
+creates a VolumeAttachment object, and if for some reason we are doing a detach
+with the in-tree plugin, the VolumeAttachment object becomes orphaned.
+
+#### New Plugin Interface
+```
+// MigratableVolumePlugin is an extended interface of VolumePlugin and is used for migration to CSI support.
+type MigratableVolumePlugin interface {
+    VolumePlugin
+    // TranslateToCSISourceJSON translates the VolumePlugin spec to a
+    // CSI Source in a JSON format that the CSI Driver can then parse
+    // - spec: the v1.Volume spec
+    TranslateToCSISourceJSON(spec *Spec) (string, error)
+    // IsMigrationEnabled returns a boolean value representing whether the in-tree
+    // plugin is migrated and should use the CSI Plugin
+    IsMigrationEnabled() bool
+}
+```
+
+#### New Attachdetach-controller Role
+```
+addControllerRole(&controllerRoles, &controllerRoleBindings, func() rbac.ClusterRole {
+        role := rbac.ClusterRole{
+            ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "attachdetach-controller"},
+            Rules: []rbac.PolicyRule{
+                …
+rbac.NewRule("patch").Groups(legacyGroup).Resources("persistentvolumes").RuleOrDie(),
+           ...
+}
+```
+
+### In-Line Volumes
+In-line controller calls are a special case because there is no PV to put an
+annotation in. In this case we will add the CSI Source JSON to the
+VolumeToAttach object and in Attach we will put the Source in a new field in the
+VolumeAttachment object VolumeAttachment.Spec.Source.InlineVolumeSource. The CSI
+Attacher will have to be modified to also check this location for a source
```

Comment:
Needs to be clarified when this `VolumeAttachment` will be created. I assume this will happen whenever an attach request is made and feature flag is enabled?

Reply:
That's correct, It will just call the CSI Attach which creates the `VolumeAttachment`, there is no special code being executed here, we are just redirecting in-tree Attach to CSI Attach instead. `VolumeAttachment` is just an implementation detail of CSI that does not exist elsewhere.